akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala

```diff
@@ -0,0 +1,299 @@
+/**
+  * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+  */
+
+package akka.http.impl.engine.client
+
+import akka.http.impl.engine.client.RedirectSupportStage.RedirectMapper
+import akka.http.scaladsl.model._
+import akka.http.scaladsl.model.headers.Location
+import akka.http.scaladsl.settings.ClientAutoRedirectSettings
+import akka.http.scaladsl.settings.ClientAutoRedirectSettings.HeadersForwardMode.{All, Only, Zero}
+import akka.stream._
+import akka.stream.scaladsl.BidiFlow
+import akka.stream.stage.{GraphStage, GraphStageLogic, InHandler, OutHandler}
+
+import scala.collection.immutable
+import scala.language.higherKinds
+import scala.util.control.NoStackTrace
+
+object RedirectSupportStage {
+
+  type RedirectMapper = (ClientAutoRedirectSettings, HttpRequest, HttpResponse) => Option[HttpRequest]
+
+  case class DefaultRedirectMapper() extends RedirectMapper {
+
+    def inferMethod(req: HttpRequest, resp: HttpResponse): Option[HttpMethod] = {
+      import HttpMethods._
+      import StatusCodes._
+      val m = req.method
+      resp.status match {
+        case MovedPermanently | Found | TemporaryRedirect | PermanentRedirect if m == GET || m == HEAD => Some(m)
+        case SeeOther => Some(GET)
+        case _ => None
+      }
+    }
+
+    def sameOrigin(u1: Uri, u2: Uri): Boolean = {
+      u2.isRelative || u1.isAbsolute && u1.authority == u2.authority
+    }
+
+    def apply(s: ClientAutoRedirectSettings, req: HttpRequest, resp: HttpResponse): Option[HttpRequest] = {
+      for {
+        location <- resp.header[Location] if resp.status.isRedirection
```

Comment:
@2beaucoup In case the `if` is outside the `for` comprehension, I'd have to step out from the `Option` monad and provide the `else` block. I don't think it looks better. Here, have a look:

      if (resp.status.isRedirection) {
        for {
          location <- resp.header[Location]
          method <- inferMethod(req, resp)
          item = if (sameOrigin(req.uri, location.uri)) s.sameOrigin else s.crossOrigin
          if item.allow
        } yield {
          val headers = item.forwardHeaders match {
            case All => req.headers
            case Zero => Nil
            case Only(hs) => req.headers.filter(h => hs.exists(h.is))
          }
          req.withUri(location.value).withHeaders(headers).withMethod(method)
        }
      }
      else {
        None
      }
    }

Instead of:

      for {
        location <- resp.header[Location] if resp.status.isRedirection
        method <- inferMethod(req, resp)
        item = if (sameOrigin(req.uri, location.uri)) s.sameOrigin else s.crossOrigin
        if item.allow
      } yield {
        val headers = item.forwardHeaders match {
          case All => req.headers
          case Zero => Nil
          case Only(hs) => req.headers.filter(h => hs.exists(h.is))
        }
        req.withUri(location.value).withHeaders(headers).withMethod(method)
      }

I'd rather keep the existing version in this case. But as you wish. Let me know if you still prefer to have the `if` outside of the `for`.